### ISSUE TITLE
sample_frac() performance issue on grouped df 

### ISSUE THREAD

**Issue by cturbelin** [tidyverse/dplyr#3229]
Hi !

I found a big difference of performance  using sample_frac() on grouped_df comparing dplyr 0.5.0 and 0.7.x

Running the following example using R/3.4.1 on RHEL 7 x64

version | mean | min | max
-------- | ------ | ---- | -----
dplyr 0.5.0 | 18 | 14 | 108
dplyr 0.7.4 | 2516 | 2384 | 3082

Run on the same machine, with specific user R library
dplyr 0.5.0 is using rlang 0.1.1, Rcpp 0.12.12
dplyr 0.7.4 is using rlang 0.1.4 and Rccp 0.12.13
Others packages are the same

I obtained the same kind of results on Windows (R.3.3 MRO, all the same except dplyr version)

```r
library(microbenchmark)
library(dplyr)

data = expand.grid(med=1:10, yw=1:36, tranche=1:4, id_tel=1:5)

do_calc_sample = function(data) {
        rr = data %>% group_by(yw, id_tel, tranche) %>% sample_frac(size = .6, replace=FALSE)
}

print(microbenchmark(do_calc_sample(data)))

```

The difference exists without grouped df, the same example removing grouping is about 4x slower with 0.7.4.

Best

**Comment by lionel-** [comment]
Thanks, we're working on it. Capture of expressions is now fast in the dev version of rlang and I'm working on making the evaluation of quosures more efficient (which is the problem here). 

`sample_frac.grouped_df()` will also need a bit of refactoring to avoid reconstructing the overscope ngroups times.

**Comment by hadley** [comment]
See also #3193

**Comment by cturbelin** [comment]
Thank you ! I've just seen the pull request

**Comment by krlmlr** [comment]
Closed in #3193.
